Zend/zend_inheritance.c

```diff
@@ -594,38 +628,78 @@ static ZEND_COLD zend_string *zend_get_function_declaration(const zend_function
 					}
 					if (precv && precv->opcode == ZEND_RECV_INIT && precv->op2_type != IS_UNUSED) {
 						zval *zv = RT_CONSTANT(precv, precv->op2);
-
-						if (Z_TYPE_P(zv) == IS_FALSE) {
-							smart_str_appends(&str, "false");
-						} else if (Z_TYPE_P(zv) == IS_TRUE) {
-							smart_str_appends(&str, "true");
-						} else if (Z_TYPE_P(zv) == IS_NULL) {
-							smart_str_appends(&str, "NULL");
-						} else if (Z_TYPE_P(zv) == IS_STRING) {
-							smart_str_appendc(&str, '\'');
-							smart_str_appendl(&str, Z_STRVAL_P(zv), MIN(Z_STRLEN_P(zv), 10));
-							if (Z_STRLEN_P(zv) > 10) {
-								smart_str_appends(&str, "...");
+						append_parameter_default_value(&str, zv);
+					}
+				} else {
+					zend_internal_arg_info *internal_arg_info = (zend_internal_arg_info*) arg_info;
+					if (internal_arg_info->default_value) {
```

Comment:
Rather than parse the default value and convert it back to a string again, can't we just directly append the string representation here?

Reply:
Yes, that would make things much easier :) but it's only suitable if we don't want to keep the behaviour in 100% sync with userland functions: for `false`, `true`, numbers and named constants it won't cause any discrepancies, but in case of `null`s, strings, expressions and arrays, we would *slightly* diverge from the original behaviour.

As far as I understood, this part of the codebase is not performance-sensitive, so maybe it's not a big problem to make some extra CPU work in order to have a consistent message? What about unifying the 2 parse implementations for making the code more readable at least? Or it doesn't matter that much?

Reply:
I think it might make more sense to go the other way around here: Represent default values for userland functions the same way as for internal functions. This would also allow us to make ReflectionParameter::getDefaultConstantName() work properly despite compile-time evaluation, and more generally allow retrieving the original form of the default value even if it is a complex expression. (We can still fetch from the RECV_INIT opcode to get the actual value.)